### ISSUE TITLE
[bug] data-layer mkdir leaks FileExistsError / FileNotFoundError on shared FS with concurrent ranks

### ISSUE THREAD

**Issue by janbernloehr** [NVIDIA-NeMo/Megatron-Bridge#3886]
## Problem

Two mkdir call sites in the SFT data layer race when many ranks call them concurrently against a shared (Lustre/NFS) filesystem:

1. `megatron.bridge.data.datasets.sft.get_dataset_root` — every caller does `Path(NEMO_DATASETS_CACHE / name).mkdir(parents=True, exist_ok=True)`.
2. `megatron.bridge.data.builders.finetuning_dataset.FinetuningDatasetBuilder.default_pack_path` — `if not p.exists(): p.mkdir(parents=True, exist_ok=True)` for the packed parquet cache directory `<dataset_root>/packed/<tokenizer>_pad_seq_to_mult<N>`. Even though `build()` gates `prepare_data()` to rank 0 with a `torch.distributed.barrier()` afterwards, every rank then re-enters `default_pack_path` via the `train_path_packed` / `validation_path_packed` / `pack_metadata` property accesses in `_build_datasets`. On Lustre, stale negative stat caching can make `default_pack_path.exists()` return False on non-rank-0 ranks even after the barrier, racing them into the same mkdir bug as site 1.

In both cases, with 32 ranks (Llama-3 70B PEFT-SFT, 8 nodes × 4 GPUs), pathlib's recursive parent-creation chain leaks `FileExistsError` / `FileNotFoundError` from one of the ancestor `os.mkdir` retries — even though the user passed `exist_ok=True` on the leaf — because the `is_dir()` swallow inside pathlib is not robust to a network-FS stat cache that lags slightly behind a peer node's just-created directory. 

## Affected area

`area:data`

## Proposed patch

```python
# src/megatron/bridge/data/datasets/sft.py
def get_dataset_root(name: str) -> Path:
    output = Path(NEMO_DATASETS_CACHE) / name
    try:
        output.mkdir(parents=True, exist_ok=True)
    except (FileExistsError, FileNotFoundError):
        pass
    return output

# src/megatron/bridge/data/builders/finetuning_dataset.py
@property
def default_pack_path(self) -> Path:
    tokenizer_model_name = self._extract_tokenizer_model_name()
    default_pack_path = (
        self.dataset_root / 'packed'
        / f'{tokenizer_model_name}_pad_seq_to_mult{self._pad_seq_to_mult}'
    )
    if not default_pack_path.exists():
        try:
            default_pack_path.mkdir(parents=True, exist_ok=True)
        except (FileExistsError, FileNotFoundError):
            pass
        logger.info(f'Using default path for packing files: {str(default_pack_path)}')
    return default_pack_path
```

Validated internally by patching our nemo container: 8-rank LoRA workloads (previously failing) now run 5 training iterations end-to-end with decreasing loss; 32-rank SFT workloads (previously failing) successfully complete dataset bootstrap and load the packed parquet on all 32 ranks.

## Environment

- Container: NVIDIA internal (PyTorch 25.x, Python 3.12, NeMo `main`, Megatron-Bridge `main` via `NeMo[common]`).
- GPU: GB200 NVL 8x4
- Workload: Llama-3 70B SFT (TP=2 PP=4) and LoRA (TP=1 PP=4), 32-rank rendezvous via NCCL.

**Comment by yaoyu-33** [comment]
Thanks for contribution, LGTM

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #3888 - [data] fix: tolerate shared filesystem mkdir races (merged 2026-05-20)
- **Issue**: #3886 - data-layer mkdir leaks FileExistsError / FileNotFoundError on shared FS with concurrent ranks

### 2. Root Cause
Two SFT data-layer call sites (`get_dataset_root` in `data/datasets/sft.py` and `FinetuningDatasetBuilder.default_pack_path` in `data/builders/finetuning_dataset.py`) performed `mkdir(parents=True, exist_ok=True)` concurrently from many ranks against shared Lustre/NFS storage; pathlib's recursive parent-creation chain can still leak `FileExistsError`/`FileNotFoundError` under concurrent creation and stale negative stat caching (non-rank-0 ranks re-enter `default_pack_path` after the barrier via packed-path property accesses). PR #3888 makes both sites tolerant of these races.

### 3. Trigger Configuration
Multi-rank SFT/PEFT dataset preparation on shared filesystems (observed at 32 ranks, Llama-3 70B PEFT-SFT, 8 nodes).

### 4. Nature of the Bug
**Classification**: CODE BUG - concurrency/filesystem tolerance (unit-level fix)

### 5. Existing Test Coverage
**In Fix PR**: YES - `tests/unit_tests/data/builders/test_finetuning_dataset.py` and `tests/unit_tests/data/datasets/test_sft.py` assert the race-tolerant behavior of both fixed call sites (exact unit tests for a unit-level defect).
**In NMFW Tests**: implicit at-scale exercise - multi-node SLURM SFT/PEFT tests (e.g. GLM-4.5V SFT/PEFT at 64 GPUs, Nemotron-3 SFT at 16-64 GPUs) run concurrent-rank dataset preparation against Lustre on every pass.

### 6. Coverage Assessment
| Test | What it covers | Covers this bug? |
|-----------|--------|------------|
| `tests/unit_tests/data/{builders,datasets}/...` (fix PR) | race-tolerant mkdir behavior at both fixed sites | Yes (exact unit tests, unit-level defect) |
| NMFW multi-node SLURM SFT tests | concurrent-rank data prep on shared Lustre at scale | Yes (crash-class — trigger path on every run) |

### 7. New Regression Test
Not needed - unit-level defect with exact fix-PR unit tests, plus at-scale exercise in existing multi-node suite runs.

### 8. Conclusion
**Verdict**: ADEQUATE COVERAGE
